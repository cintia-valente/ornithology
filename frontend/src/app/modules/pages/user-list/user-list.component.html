<section class="user-list">
  <div *ngFor="let user of users">
    <div class="user-name">
      {{ user.name }}
    </div>
    <div class="user-email">
      {{ user.email }}
    </div>
    <div class="user-password">
      {{ user.password }}
    </div>

    <div class="delete">
      <button type="submit" class="btn btn-primary" (click)="deleteUser(user)">
        Excluir
      </button>
    </div>
  </div>
</section>

<!-- <ng-template #loading>
  <div class="error" *ngIf="error">
    Erro ao carregar usu√°rio. Tente novamente mais tarde.
  </div>
</ng-template> -->
