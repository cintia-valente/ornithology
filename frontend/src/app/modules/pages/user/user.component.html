<div class="users">
  <div *ngIf="!error; else loading">
    <div *ngFor="let user of users">
      <div class="user-name">
        {{ user.name }}
      </div>
      <div class="user-email">
        {{ user.email }}
      </div>
      <div class="user-password">
        {{ user.password }}
      </div>
    </div>
  </div>

  <ng-template #loading>
    <div class="error" *ngIf="error">
      Erro ao carregar usu√°rio. Tente novamente mais tarde.
    </div>
  </ng-template>

  <div class="register">
    <button type="submit" class="btn btn-primary">Atualizar</button>
  </div>
</div>
