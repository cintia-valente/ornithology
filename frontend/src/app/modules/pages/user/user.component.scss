.error {
    color: #b70000;
}