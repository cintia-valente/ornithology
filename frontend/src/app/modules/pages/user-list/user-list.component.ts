import { UserService } from './../../../services/user.service';
import { Component, OnInit } from '@angular/core';
import { User } from 'src/app/model/user.model';
import { HttpErrorResponse } from '@angular/common/http';
import { throwError } from 'rxjs';
import { ToastrService } from 'ngx-toastr';
@Component({
  selector: 'app-user-list',
  templateUrl: './user-list.component.html',
  styleUrls: ['./user-list.component.scss'],
})
export class UserListComponent implements OnInit {
  public users: User[] = [];
  error: boolean = false;
  submmited: boolean = false;

  constructor(
    private userService: UserService,
    private toastr: ToastrService
  ) {}

  ngOnInit(): void {
    this.listUser();
  }

  listUser() {
    this.error = false;

    this.userService.getUsers().subscribe({
      next: (data: User[]) => (this.users = data),

      error: (err: HttpErrorResponse) => {
        this.toastr.error(
          'Erro ao carregar usuÃ¡rios. Por favor, tente novamente mais tarde'
        );

        return throwError(() => err);
      },
    });
  }

  deleteUser(user: User) {
    this.userService.deleteUser(user.idUser).subscribe({
      next: () => {
        this.listUser();
        this.toastr.error('ExcluÃ­do com sucesso.');
      },
      error: (err: HttpErrorResponse) => {
        this.toastr.error(
          'Erro ao excluir usuÃ¡rio. Por favor, tente novamente mais tarde.'
        );

        return throwError(() => err);
      },
    });
  }
}
