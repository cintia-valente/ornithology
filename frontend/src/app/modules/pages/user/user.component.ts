import { UserService } from './../../../services/user.service';
import { Component, OnInit } from '@angular/core';
import { User } from 'src/app/model/user.model';
import { HttpErrorResponse } from '@angular/common/http';
import { throwError } from 'rxjs';
@Component({
  selector: 'app-user',
  templateUrl: './user.component.html',
  styleUrls: ['./user.component.scss'],
})
export class UserComponent implements OnInit {
  public users: User[] = [];
  error: boolean = false;
  loading: boolean = false;
  isEdit: boolean = false;
  submmited: boolean = false;

  constructor(private userService: UserService) {}

  ngOnInit(): void {
    this.listUser();
  }

  listUser() {
    this.error = false;

    this.userService.getUsers().subscribe({
      next: (data: User[]) => console.log((this.users = data)),

      error: (err: HttpErrorResponse) => {
        this.error = true;

        return throwError(() => err);
      },
    });
  }

  deleteUser(user: User) {
    this.userService.deleteUser(user.idUser).subscribe(() => {
      alert(`Usuário excluído com sucesso!`);
      this.listUser();
    });
  }
}
