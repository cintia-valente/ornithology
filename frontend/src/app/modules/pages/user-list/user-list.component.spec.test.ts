import { CommonModule } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import {
  ComponentFixture,
  getTestBed,
  TestBed,
  TestComponentRenderer,
} from '@angular/core/testing';
import { RouterTestingModule } from '@angular/router/testing';
import { Routes } from '@angular/router';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';

import { ToastrService } from 'ngx-toastr';
import { of, throwError } from 'rxjs';

import { UserService } from '../../../services/user.service';

import { UserListComponent } from './user-list.component';

const mockUsers = [
  {
    idUser: '135',
    name: 'Maite',
    email: 'maite@email.com',
    password: '',
  },
];

class UserServiceMock {
  getUsers() {
    return of();
  }

  deleteUser(id: any) {
    return of();
  }
}

class ToastrServiceMock {
  success(message?: string) {}

  error(message?: string) {}
}

const routes: Routes = [
  {
    path: 'editar-anotacoes/:id',
    component: TestComponentRenderer,
  },
];

describe('UserListComponent', () => {
  let injector: TestBed;
  let component: UserListComponent;
  let fixture: ComponentFixture<UserListComponent>;
  let userService: UserService;
  let toastrService: ToastrService;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [
        CommonModule,
        HttpClientModule,
        RouterTestingModule,
        RouterTestingModule.withRoutes(routes),
      ],
      declarations: [UserListComponent],
      schemas: [CUSTOM_ELEMENTS_SCHEMA],
      providers: [
        { provide: UserService, useClass: UserServiceMock },
        { provide: ToastrService, useClass: ToastrServiceMock },
      ],
    }).compileComponents();
  });

  beforeEach(() => {
    injector = getTestBed();
    fixture = TestBed.createComponent(UserListComponent);
    //location = injector.inject(Location);

    userService = injector.inject(UserService);
    toastrService = injector.inject(ToastrService);

    jest.spyOn(userService, 'getUsers').mockReturnValue(of(mockUsers));

    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it(`Dado: que o componente foi carregado
      Então: deve chamar o serviço userService.getUsers`, async () => {
    //Arrange
    const spyUserList = jest
      .spyOn(userService, 'getUsers')
      .mockReturnValue(of(mockUsers));

    //Act
    fixture.detectChanges();

    //Assert
    expect(spyUserList).toHaveBeenCalled();
  });

  it(`Dado: que o componente foi carregado
      Quando: ocorrer um erro ao carregar os birds
      Então: deve chamar o serviço userService.getUsers`, async () => {
    //Arrange
    const error = {
      error: 'error',
      status: 400,
      message: 'Server Error',
    };

    const spyUserListError = jest
      .spyOn(userService, 'getUsers')
      .mockReturnValue(throwError(() => error.message));

    //Act
    fixture.detectChanges();

    //Assert
    expect(spyUserListError).toHaveBeenCalled();
  });

  it(`Dado: que o componente foi carregado
      E: clicar no ícone de deletar
      Então: deve chamar o serviço userService.deleteUser`, async () => {
    //Arrange
    const mockUser = {
      idUser: '135',
      name: 'Maite',
      email: 'maite@email.com',
      password: '',
    };

    const spyUserDelete = jest
      .spyOn(userService, 'deleteUser')
      .mockReturnValue(of(mockUser));

    fixture.detectChanges();
    await fixture.whenStable();

    const userDelete = fixture.nativeElement.querySelector(
      '.users table > tbody > tr:nth-child(1) > td:nth-child(3) > div > span.icon-delete > img'
    );

    userDelete.click();

    fixture.detectChanges();
    await fixture.whenStable();

    //Assert
    expect(spyUserDelete).toHaveBeenCalled();
  });

  it(`Dado: que o componente foi carregado
      Quando: atribuir um erro no serviço 
      Então: deve chamar o serviço userService.deleteUser`, async () => {
    //Arrange
    const error = {
      error: 'error',
      status: 400,
      message: 'Server Error',
    };

    const spyUserDeleteError = jest
      .spyOn(userService, 'deleteUser')
      .mockReturnValue(throwError(() => error.message));

    //Act
    fixture.detectChanges();
    await fixture.whenStable();

    const userDelete = fixture.nativeElement.querySelector(
      '.users table > tbody > tr:nth-child(1) > td:nth-child(3) > div > span.icon-delete > img'
    );

    userDelete.click();

    fixture.detectChanges();
    await fixture.whenStable();

    //Assert
    expect(spyUserDeleteError).toHaveBeenCalled();
  });

  it(`Dado: que o componente foi carregado
      Quando: detectar as alterações
      Então: deve gerar um Snapshot do componente carregado`, async () => {
    //Arrange
    jest.spyOn(userService, 'getUsers').mockReturnValue(of(mockUsers));

    //Act
    fixture.detectChanges();

    //Assert
    expect(fixture).toMatchSnapshot();
  });
});
