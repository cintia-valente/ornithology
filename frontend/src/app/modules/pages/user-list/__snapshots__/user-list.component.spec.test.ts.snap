// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserListComponent Dado: que o componente foi carregado
      Quando: detectar as alterações
      Então: deve gerar um Snapshot do componente carregado 1`] = `
<app-user-list
  __ngContext__={[Function Number]}
  error="false"
  submmited="false"
  toastr={[Function ToastrServiceMock]}
  userService={[Function UserServiceMock]}
  users={[Function Array]}
>
  <app-nav-bar /><div
    class="users"
  >
    <div
      class="container"
    >
      <div
        class="row justify-content-center"
      >
        <div
          class="col-xl-8"
        >
          <div
            class="section-title"
          >
            Usuários
          </div>
          <table
            class="table table-striped"
          >
            <thead>
              <tr>
                <th>
                  Nome
                </th>
                <th>
                  Email
                </th>
                <th>
                  Ações
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td
                  class="py-1"
                >
                   Maite 
                </td>
                <td
                  class="py-1"
                >
                   maite@email.com 
                </td>
                <td
                  class="py-1"
                >
                  <div
                    class="table-actions"
                  >
                    <span
                      class="icon-delete"
                    >
                      <img
                        src="assets/svg/layout/trash-solid.svg"
                      />
                    </span>
                    <span
                      class="icon-edit"
                    >
                      <a
                        href="/editar-usuario/135"
                      >
                        <img
                          src="assets/svg/layout/pen-solid.svg"
                          style="width: 17px;"
                        />
                      </a>
                    </span>
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</app-user-list>
`;
